### PR TITLE
added unsafe memcpy bytes to/from double

### PR DESCRIPTION
**Pull request by jbloom22**
- used in read/write of breeze dense matrix

### CODE REVIEW

**Comment by jbloom22** [comment]
This builds on the changes in #2494 incorporating transpose, I'll rebase whichever goes in second.

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala [hunk 1]
```diff
@@ -14,18 +15,22 @@ object RichDenseMatrixDouble {
...
+    val length = rows * cols
     
-    val data = new Array[Double](rows * cols)
-    var i = 0
-    while (i < rows * cols) {
-      data(i) = in.readDouble()
-      i += 1
-    }
+    val n = length << 3
+    val bytes = Array.ofDim[Byte](n)
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: `ofDim` annoys me, just call the `Array` construtor: `new Array[Byte](n)`, etc.

**Comment by cseed** [comment]
Compare: https://github.com/hail-is/hail/pull/2526#issuecomment-349842098

```
('write A', [1.2455670833587646, 0.7725400924682617, 0.6804611682891846])
('write B', [4.475854158401489, 4.356597900390625, 3.8925211429595947])
('mul', [42.79294395446777, 43.01479911804199, 41.07988786697388])
```

Seemed to speed up the IO quite a bit (30%?) the matmul less so.

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala (hunk 1 above)
Comment (cseed) [reply src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: You missed this one.

**Comment by jbloom22** [comment]
thanks for catching that third one, should be good now

**Comment by jbloom22** [comment]
looking into test failure

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala [hunk 2]
```diff
@@ -19,17 +20,17 @@ object RichDenseMatrixDouble {
...
     
-    val data = new Array[Double](rows * cols)
-    var i = 0
-    while (i < rows * cols) {
-      data(i) = in.readDouble()
-      i += 1
-    }
+    val n = length << 3
+    val bytes = new Array[Byte](n)
+    in.read(bytes, 0, n)
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: This needs to be `readFully`.  This is probably your bug.

Comment (jbloom22) [reply src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: Ha, I started writing a loop for reading fully but your note that it exists. I think the issue may be that writeDouble on DataOutputStream first converts to a long representation using doubleToLongBits, which may not be the raw bytes. I am going to change the IRM to Block export to use Array[Byte] as well.

I also realized that using an Array[Byte] divides by 8 the largest dimension of local matrix we can output, from 46k to 5792. Unless you have a better idea, I'll fix that by putting the write in a loop that takes up to maxInt bytes at a time.

**Comment by jbloom22** [comment]
Yes, please re-review. I started to add the while loops in order to handle case where array length is greater than maxInt / 8, but then decided to make that a separate PR.

**Comment by jbloom22** [comment]
I added 64KB buffer, but I'm unsure of the right interface (should the caller pass in the byte buffer as well?) and location (RichDenseMatrixDouble is quite specific) for these functions.

In compute on WriteBlocksRDD, I create a byte buffer of the same size in bytes as the entire IndexedRow, and re-use it for every row. This limits the number of cols to 268 million or so (~2^28) which I don't see us exceeding anytime soon. I could instead use the writeDoubles function but then I'd want to change the interface to pass in the byte buffer so it's not recreated for every row. What do you think is most reasonable?

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala [hunk 3]
```diff
@@ -14,22 +15,55 @@ object RichDenseMatrixDouble {
       Some(DenseMatrix.horzcat(ms: _*))
   }
   
+  // copies n doubles as bytes from data to dos
+  def writeDoubles(dos: DataOutputStream, data: Array[Double], n: Int) {
+    assert(n <= data.length)
+    var nLeft = n
+    val bufSize = if (nLeft < (1 << 14)) nLeft else 1 << 14 // up to 8k doubles
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: math.min(nLeft, 1 << 14)

1 << 14 is 16K, FYI.  Might be clearer to write this as 8 * 1024, say (8192 is also fine in my book.)

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala [hunk 4]
```diff
@@ -14,22 +15,55 @@ object RichDenseMatrixDouble {
...
+    val buf = new Array[Byte](bufSize << 3) // up to 64k bytes
+
+    var off = 0
+    while (nLeft > 0) {
+      val nCopy = if (nLeft <= bufSize) nLeft else bufSize
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: `math.min(nLeft, bufSize)`

File: src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala [hunk 5]
```diff
@@ -14,22 +15,55 @@ object RichDenseMatrixDouble {
...
+      nLeft -= nCopy
+    }
+  }
+
+  // copies n doubles as bytes from dis to data
+  def readDoubles(dis: DataInputStream, data: Array[Double], n: Int) {
+    assert(n <= data.length)
+    var nLeft = n
+    val bufSize = if (nLeft < (1 << 14)) nLeft else 1 << 14 // up to 8k doubles
+    val buf = new Array[Byte](bufSize << 3) // up to 64k bytes
```
Comment (cseed) [inline src/main/scala/is/hail/utils/richUtils/RichDenseMatrixDouble.scala]: Same comments as above.

**Pull request merged**